### PR TITLE
Number of accounts option added in create account command - Closes #423

### LINKED ISSUE #423: Add option to create multiple accounts
### Description

Sometimes it would be helpful to create multiple accounts in one go. E.g. if setting up a third-party application which needs to distinguish between payments from different sources.

**Expected behavior:** `create account -n 50` would create 50 accounts and return them in an array

**Actual behavior:** Only one account can be created at a time

### PR DESCRIPTION
**Pull request by Mastermaulik**
### What was the problem?
The create account command didn't have the option to specify the number of accounts to create.

### How did I fix it?
Added an option to get the number of accounts to be created in the create account command.

### How to test it?
create account -n 10

### Review checklist

* The PR solves #423 
* All new code is covered with unit tests
* All new code was formatted with Prettier
* Linting passes
* Tests pass
* Commit messages follow the [commit guidelines](CONTRIBUTING.md#git-commit-messages)
* Documentation has been added/updated

### CODE REVIEW

**Comment by shuse2** [comment]
Thank you for the PR.
Please add the test regards to this PR, too.

File: src/commands/create_account.js [hunk 1]
```diff
@@ -16,29 +16,40 @@
...
-		privateKey,
-		publicKey,
-		address,
-	};
+export const actionCreator = () => async ({ options = {} }) => {
+	const { number = 1 } = options || {};
+
+	const accountsArray = [];
+
+	for (let i = 0; i < number; i += 1) {
```
Comment (shuse2) [inline src/commands/create_account.js]: Prefer not using `for-loop`

File: src/utils/options.js [hunk 2]
```diff
@@ -69,10 +69,14 @@ const unvotesDescription = `Specifies the public keys for the delegate candidate
 const noSignatureDescription =
 	'Creates the transaction without a signature. Your passphrase will therefore not be required.';
 
+const numOfAccountsDescription =
```
Comment (shuse2) [inline src/utils/options.js]: prefer verbose name `num` => `number`

File: test/steps/crypto/3_then.js [hunk 3]
```diff
@@ -121,12 +121,14 @@ export function itShouldResolveToAnObjectWithThePassphraseThePrivateKeyThePublic
 		keys: { privateKey, publicKey },
 		address,
 	} = this.test.ctx;
-	const expectedObject = {
-		passphrase,
-		privateKey,
-		publicKey,
-		address,
-	};
+	const expectedObject = [
```
Comment (shuse2) [inline test/steps/crypto/3_then.js]: probably this change is not intended?

Comment (Mastermaulik) [reply test/steps/crypto/3_then.js]: This change is intended as the returned value from the function is going to be an array, earlier the returned value was an object and thus the expected object would change from being an object to an array.

File: src/commands/create_account.js (hunk 1 above)
Comment (shuse2) [reply src/commands/create_account.js]: u don't need lodash to loop through the number.
it's like
```
new Array(10).fill(0).map(i => 'log_i');
```

File: test/steps/crypto/3_then.js (hunk 3 above)
Comment (shuse2) [reply test/steps/crypto/3_then.js]: ah yes, but probably better to add multiple elements check too

Comment (Mastermaulik) [reply test/steps/crypto/3_then.js]: @shuse2 , can you please tell me an example of existing code where we have tests for arrays?

The current test was going to just check for the object.

I had some doubts related to the regex that we are using to get the variables.

```
'a passphrase "minute omit local rare sword knee banner pair rib museum shadow juice" with private key "314852d7afb0d4c283692fef8a2cb40e30c7a5df2ed79994178c10ac168d6d977ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and public key "7ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and address "2167422481642255385L"'
```

How do we convert this given statement to use an array?

Comment (Mastermaulik) [reply test/steps/crypto/3_then.js]: @willclarktech , can you please guide me on how to add tests for an array?

Comment (shuse2) [reply test/steps/crypto/3_then.js]: I think it is worth to look at how `list` command works.
Also, what I mean by test array, some edge cases like `n <= 0`.
Also, checking, both single and multiple results are as expected

File: src/commands/create_account.js [hunk 4]
```diff
@@ -14,31 +14,55 @@
...
+export const actionCreator = () => async ({ options }) => {
+	const { number = 1 } = options;
+
+	if (!checkIsInt(number)) {
+		throw new ValidationError(
+			`${number} doesn't represent a valid Integer corresponding to number of accounts`,
+		);
+	}
+
+	if (!checkIsNonZeroInt(number)) {
```
Comment (shuse2) [inline src/commands/create_account.js]: I think nonzero interger doesn't mean it is positive number

File: src/utils/options.js [hunk 5]
```diff
@@ -3,47 +3,47 @@ const jsonDescription =
 
 const messageDescription = `Specifies a source for providing a message to the command. If a string is provided directly as an argument, this option will be ignored. The message must be provided via an argument or via this option. Sources must be one of \`file\` or \`stdin\`. In the case of \`file\`, a corresponding identifier must also be provided.
 
-	Note: if both secret passphrase and message are passed via stdin, the passphrase must be the first line.
+Note: if both secret passphrase and message are passed via stdin, the passphrase must be the first line.
 
-	Examples:
-	- --message file:/path/to/my/message.txt
-	- --message stdin
+Examples:
```
Comment (shuse2) [inline src/utils/options.js]: Why was there a format change here?

File: test/steps/crypto/1_given.js [hunk 6]
```diff
@@ -93,6 +93,12 @@ export function aPassphraseWithPublicKey() {
 	this.test.ctx.publicKey = publicKey;
 }
 
+export function aStringDenotingNumberOfAccounts() {
```
Comment (shuse2) [inline test/steps/crypto/1_given.js]: `aNumberOfAccounts` seems understandable
and also is this function really relates to `crypto`?

File: test/steps/crypto/1_given.js [hunk 7]
```diff
@@ -130,6 +136,48 @@ export function aPassphraseWithPrivateKeyAndPublicKeyAndAddress() {
 	this.test.ctx.address = address;
 }
 
+export function numberOfAccountsAsOneAndPassphraseWithPrivateKeyAndPublicKeyAndAddress() {
```
Comment (shuse2) [inline test/steps/crypto/1_given.js]: numberOfAccounts seems to be taken from the title, and non-one

File: src/commands/create_account.js (hunk 4 above)
Comment (Mastermaulik) [reply src/commands/create_account.js]: I am replacing the nonzero check with checkIsPositiveInt.

File: src/utils/options.js (hunk 5 above)
Comment (Mastermaulik) [reply src/utils/options.js]: I think this is because I made changes to a temporary file and then copied the file back, I will revert these changes.

File: test/steps/crypto/1_given.js (hunk 7 above)
Comment (Mastermaulik) [reply test/steps/crypto/1_given.js]: this function's name has been changed to numberOfAccountsAndPassphraseWithPrivateKeyAndPublicKeyAndAddress.

File: test/steps/crypto/1_given.js (hunk 6 above)
Comment (Mastermaulik) [reply test/steps/crypto/1_given.js]: Changed this function's name to aNumberOfAccounts.

I made this part of crypto, as the aPassphraseWithPrivateKeyAndPublicKeyAndAddress function was part of crypto file.
I think that options/1_given is more apt for this, shall I move it there?

File: src/commands/create_account.js (hunk 4 above)
Comment (Mastermaulik) [reply src/commands/create_account.js]: It looks like: const checkIsPositiveInt = number => parseInt(number, 10) > 0;

File: src/commands/create_account.js [hunk 8]
```diff
@@ -14,31 +14,55 @@
...
+		const { address } = cryptography.getAddressFromPublicKey(publicKey);
+		return {
+			passphrase,
+			privateKey,
+			publicKey,
+			address,
+		};
+	});
+
+	return accountsArray;
```
Comment (willclarktech) [inline src/commands/create_account.js]: This would be a breaking change. I think the normal use case is probably going to be a single account. How about returning an object if the option isn't passed, and a list otherwise?

File: src/commands/create_account.js [hunk 9]
```diff
@@ -14,31 +14,55 @@
...
+const checkIsInt = number => Number.isInteger(parseInt(number, 10));
+
+const checkIsPositiveInt = number => parseInt(number, 10) > 0;
+
+export const actionCreator = () => async ({ options }) => {
+	const { number = 1 } = options || {};
+
+	if (!checkIsInt(number)) {
+		throw new ValidationError(
+			`${number} doesn't represent a valid Integer corresponding to number of accounts`,
```
Comment (willclarktech) [inline src/commands/create_account.js]: `Number of accounts must be a positive integer`

File: src/commands/create_account.js [hunk 10]
```diff
@@ -14,31 +14,55 @@
  *
  */
 import cryptography from '../utils/cryptography';
+import { ValidationError } from '../utils/error';
 import { createCommand } from '../utils/helpers';
 import { createMnemonicPassphrase } from '../utils/mnemonic';
+import commonOptions from '../utils/options';
 
-const description = `Returns a randomly-generated mnemonic passphrase with its corresponding public key and address.
+const description = `Returns a list of randomly-generated mnemonic passphrase with its corresponding public key and address.
```
Comment (willclarktech) [inline src/commands/create_account.js]: If my suggestion below is accepted, this will have to be updated too.

File: src/commands/create_account.js [hunk 11]
```diff
@@ -14,31 +14,55 @@
...
-	const { privateKey, publicKey } = cryptography.getKeys(passphrase);
-	const { address } = cryptography.getAddressFromPublicKey(publicKey);
-
-	return {
-		passphrase,
-		privateKey,
-		publicKey,
-		address,
-	};
+const checkIsInt = number => Number.isInteger(parseInt(number, 10));
```
Comment (willclarktech) [inline src/commands/create_account.js]: Prefer full words: `checkIsInteger` or even `isInteger`

File: src/commands/create_account.js [hunk 12]
```diff
@@ -14,31 +14,55 @@
...
+
+const checkIsPositiveInt = number => parseInt(number, 10) > 0;
```
Comment (willclarktech) [inline src/commands/create_account.js]: I think these checks can be combined.

File: src/commands/create_account.js [hunk 13]
```diff
@@ -14,31 +14,55 @@
...
+		throw new ValidationError(
+			`${number} doesn't represent a valid Integer corresponding to number of accounts`,
+		);
+	}
+
+	if (!checkIsPositiveInt(number)) {
+		throw new ValidationError('Please provide a number greater than zero');
+	}
+
+	const accountsArray = new Array(number).fill(0).map(() => {
```
Comment (willclarktech) [inline src/commands/create_account.js]: `0` in `.fill` is unnecessary.

File: src/utils/options.js [hunk 14]
```diff
@@ -69,10 +69,17 @@ const unvotesDescription = `Specifies the public keys for the delegate candidate
 const noSignatureDescription =
 	'Creates the transaction without a signature. Your passphrase will therefore not be required.';
 
+const numberOfAccountsDescription =
```
Comment (willclarktech) [inline src/utils/options.js]: This option is specific to the `create account` command, so probably belongs in that file rather than here. Compare with `--output-public-key` in the `encrypt passphrase` command.

File: test/specs/commands/create_account.js [hunk 15]
```diff
@@ -20,30 +20,96 @@ import * as then from '../../steps/3_then';
...
+		'a passphrase "minute omit local rare sword knee banner pair rib museum shadow juice" with private key "314852d7afb0d4c283692fef8a2cb40e30c7a5df2ed79994178c10ac168d6d977ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and public key "7ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and address "2167422481642255385L"',
+		given.aPassphraseWithPrivateKeyAndPublicKeyAndAddress,
+		() => {
+			Given(
+				'the passphrase is generated by the createMnemonicPassphrase function',
+				given.thePassphraseIsGeneratedByTheCreateMnemonicPassphraseFunction,
+				() => {
+					Given('an action "create account"', given.anAction, () => {
+						When('the action is called', when.theActionIsCalled, () => {
+							Then(
```
Comment (willclarktech) [inline test/specs/commands/create_account.js]: This test could be nested more efficiently if `Given a valid integer corresponding to number of accounts "1"` was more deeply nested.

File: test/steps/crypto/3_then.js (hunk 3 above)
Comment (willclarktech) [inline test/steps/crypto/3_then.js]: This should be renamed to `expectedArray`.

File: test/specs/commands/create_account.js [hunk 16]
```diff
@@ -20,30 +20,96 @@ import * as then from '../../steps/3_then';
...
+		},
+	);
 	Given(
 		'a crypto instance has been initialised',
 		given.aCryptoInstanceHasBeenInitialised,
 		() => {
 			Given(
-				'a passphrase "minute omit local rare sword knee banner pair rib museum shadow juice" with private key "314852d7afb0d4c283692fef8a2cb40e30c7a5df2ed79994178c10ac168d6d977ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and public key "7ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and address "2167422481642255385L"',
-				given.aPassphraseWithPrivateKeyAndPublicKeyAndAddress,
+				'a valid integer corresponding to number of accounts "1"',
```
Comment (willclarktech) [inline test/specs/commands/create_account.js]: It would be good to have tests for larger numbers too, which was after all the point of the feature.

Comment (Mastermaulik) [reply test/specs/commands/create_account.js]: @willclarktech , I was thinking about how to write the tests for large numbers.

As far as I understand the code,
we make the functions return the passphrase, keys and address by doing:
lisk.cryptography.getAddressFromPublicKey.returns(address);

Here if we think that we can create cases where we have a multiple number of accounts, how do we change the test?
We can have something like this for 2 accounts:

``` 
Given(
'a valid integer corresponding to number of accounts "2" 
 for the first account, passphrase "smart couch position escape begin retreat summer snack fruit virtual climb shine" with private key "3818764e457d57dada65de626660039307fdd3fbf83743095579188669a110a8f4175ac1892a450df668b75a90e7a3b5066d0bc90ee91a808de686a8867dc46d" and public key "f4175ac1892a450df668b75a90e7a3b5066d0bc90ee91a808de686a8867dc46d" and address "7195654441783367976L"
  and for second account,  passphrase "minute omit local rare sword knee banner pair rib museum shadow juice" with private key "314852d7afb0d4c283692fef8a2cb40e30c7a5df2ed79994178c10ac168d6d977ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and public key "7ef45cd525e95b7a86244bbd4eb4550914ad06301013958f4dd64d32ef7bc588" and address "2167422481642255385L"
'),
```	
The parsing logic can be written here, but my doubt is that how do we change `aPassphraseWithPrivateKeyAndPublicKeyAndAddress` function to return different parameters in each iteration of create account call.

Comment (willclarktech) [reply test/specs/commands/create_account.js]: My suggestion would be to have distinct steps for distinct setup requirements. So one step would specify that the number of accounts is 2, another step would specify the passphrase with corresponding keys and address, and a third step would specify another passphrase/keys/address set: `anotherPassphraseWithPrivateKeyAndPublicKeyAndAddress`.

Comment (willclarktech) [reply test/specs/commands/create_account.js]: For distinct variables stored in the test context we need distinct names. So e.g. `passphrase` and `passphrase2`, `publicKey` and `publicKey2` etc.

**Comment by shuse2** [comment]
Closing due to no response.
Also, this issue will be resolved in `v2`
